src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -297,135 +321,72 @@ public virtual void Write(long value)
+            OutStream.Write(buffer);
         }
 
         // Writes a length-prefixed string to this stream in the BinaryWriter's
         // current Encoding. This method first writes the length of the string as
         // an encoded unsigned integer with variable length, and then writes that many characters
         // to the stream.
         //
-        public virtual unsafe void Write(string value)
+        public virtual void Write(string value)
         {
             if (value == null)
                 throw new ArgumentNullException(nameof(value));
 
-            int totalBytes = _encoding.GetByteCount(value);
-            Write7BitEncodedInt(totalBytes);
-
-            if (_largeByteBuffer == null)
-            {
-                _largeByteBuffer = new byte[LargeByteBufferSize];
-                _maxChars = _largeByteBuffer.Length / _encoding.GetMaxByteCount(1);
-            }
-
-            if (totalBytes <= _largeByteBuffer.Length)
-            {
-                _encoding.GetBytes(value, _largeByteBuffer);
-                OutStream.Write(_largeByteBuffer, 0, totalBytes);
-                return;
-            }
+            // Common: UTF-8, small string, avoid 2-pass calculation
+            // Less common: UTF-8, large string, avoid 2-pass calculation
+            // Uncommon: excessively large string or not UTF-8
 
-            int numLeft = value.Length;
-            int charStart = 0;
-            ReadOnlySpan<char> str = value;
-
-            // The previous implementation had significant issues packing encoded
-            // characters efficiently into the byte buffer. This was due to the assumption,
-            // that every input character will take up the maximum possible size of a character in any given encoding,
-            // thus resulting in a lot of unused space within the byte buffer.
-            // However, in scenarios where the number of characters aligns perfectly with the buffer size the new
-            // implementation saw some performance regressions, therefore in such scenarios (ASCIIEncoding)
-            // work will be delegated to the previous implementation.
-            if (_encoding.GetType() == typeof(UTF8Encoding))
+            if (_useFastUtf8)
             {
-                while (numLeft > 0)
+                if (value.Length <= 24)
                 {
-                    _encoder.Convert(str.Slice(charStart), _largeByteBuffer, numLeft <= _maxChars, out int charCount, out int byteCount, out bool _);
-
-                    OutStream.Write(_largeByteBuffer, 0, byteCount);
-                    charStart += charCount;
-                    numLeft -= charCount;
+                    Span<byte> buffer = stackalloc byte[24 * 3]; // max expansion: each char -> 3 bytes (use const for improved perf)
+                    int actualByteCount = _encoding.GetBytes(value, buffer);
+                    Write7BitEncodedInt(actualByteCount);
```

Comment:
This is going to do its own Write(byte) to the OutStream; best case that's a virtual call that buffers, but it could also be a syscall (e.g. if OutStream were a NetworkStream).  Could we instead just allocate the buffer with a + 1 on the number of bytes, store this into the buffer[0], _encoding.GetBytes into buffer.Slice(1), and then OutStream.Write(buffer.Slice(0, actualByteCount + 1)?

Reply:
Seems reasonable. Should we also consider bumping the constant values used for this scenario? That is: `if (value.Length <= 127 / 3) { stackalloc byte[127]; /* ... */ }`.

Means we're stackallocing 127 bytes instead of 72 bytes, but... should be ok?

Reply:
> stackallocing 127 bytes

The JIT will round this to 128 bytes?